### PR TITLE
[R4R]BEP3: HTLC and Atomic Peg

### PR DESCRIPTION
**Pull request by HaoyangLiu**
This BEP is about introducing Hash Timer Locked Contract functions and further mechanism to handle inter-blockchain token peg.

[HTLC](https://en.bitcoin.it/wiki/Hash_Time_Locked_Contracts) has been used for Atomic Swap and cross payment channel for a few years on Bitcoin and its variant blockchains, and also Ethereum. This BEP defines native transactions to support HTLC on Binance Chain, and also proposes the standard infrastructure and procedure to use HTLC for inter-chain atomic swap to easily create and use pegged token, which is called `Atomic Peg`.

Related BEP: https://github.com/binance-chain/BEPs/pull/3

### CODE REVIEW

**Comment by HaoyangLiu** [comment]
@caffeinum Could you help to review this BEP?

File: BEP3.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+Hash Timer Locked Transfer (HTLT) is a new transaction type on Binance Chain, to serve as HTLC in the first step of Atomic Swap, with parameters defined as the below:
+
+| Name | Type | Description | Optional |
+| -----| ---- | ----------- | -------- |
+| From | Address | Sender address, where the asset is from | No| 
+| To | Address | Receiver address, where the asset is to, if the proper condition meets. This address must be flagged with HTLT flag in order to claim the transfer to be fully done once the hashed secret is disclosed by Sender. | No | 
+| ToOnOtherChain  | bytes   | an byte array, maximum 32 bytes, in any proper encoding | No            | 
+| Random Number Hash | 32 bytes   | Sender address, where the asset is from | No | 
+| Coins | []Coin   |similar to the Coins in the original Transfer defined in BEP2| No | 
+| TimeSpan | int64   | number of blocks to wait before the asset may be returned to From if not claimed via Random. The number must be larger than or equal to 360 (>2 minutes), and smaller than 518400 (>48 hours) | No |
```
Comment (yutianwu) [inline BEP3.md]: should be `<48 hours`

File: BEP3.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+### Hash Timer Locked Transfer
+
+Hash Timer Locked Transfer (HTLT) is a new transaction type on Binance Chain, to serve as HTLC in the first step of Atomic Swap, with parameters defined as the below:
+
+| Name | Type | Description | Optional |
+| -----| ---- | ----------- | -------- |
+| From | Address | Sender address, where the asset is from | No| 
+| To | Address | Receiver address, where the asset is to, if the proper condition meets. This address must be flagged with HTLT flag in order to claim the transfer to be fully done once the hashed secret is disclosed by Sender. | No | 
+| ToOnOtherChain  | bytes   | an byte array, maximum 32 bytes, in any proper encoding | No            |
```
Comment (yutianwu) [inline BEP3.md]: `ToOnOtherChain`, is this a typo

File: BEP3.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+### Hash Timer Locked Transfer
+
+Hash Timer Locked Transfer (HTLT) is a new transaction type on Binance Chain, to serve as HTLC in the first step of Atomic Swap, with parameters defined as the below:
+
+| Name | Type | Description | Optional |
+| -----| ---- | ----------- | -------- |
+| From | Address | Sender address, where the asset is from | No| 
+| To | Address | Receiver address, where the asset is to, if the proper condition meets. This address must be flagged with HTLT flag in order to claim the transfer to be fully done once the hashed secret is disclosed by Sender. | No | 
+| ToOnOtherChain  | bytes   | an byte array, maximum 32 bytes, in any proper encoding | No            | 
+| Random Number Hash | 32 bytes   | Sender address, where the asset is from | No |
```
Comment (yutianwu) [inline BEP3.md]: comment should be wrong

File: BEP3.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+
+### Claim Hash Timer Locked 
+
+Claim Hash Timer Locked (CHTL) is to claim the locked asset by showing the Random Number value that matches the hash. Each HTLT locked asset is guaranteed to be release once.
+
+| Name | Type | Description | Optional |
+| -----| ---- | ----------- | -------- |
+| From | Address | Sender address, where the asset is from | No| 
+| Random Number Hash | 32 bytes   | Sender address, where the asset is from | No |
```
Comment (yutianwu) [inline BEP3.md]: comment is wrong

File: BEP3.md (hunk 2 above)
Comment (HaoyangLiu) [reply BEP3.md]: `ToOnOtherChain` represents the expected receiver address on the other blockchain. Different blockchain may adopt different address encoding methods. So here we use byte array type.

**Comment by drunken005** [comment]
When will this feature come online

File: BEP3.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+*   **Client tooling**: wallets or other tools that help Clients to monitor the blockchains to complete the claim (and maybe the lock as well).
+
+### Protocol
+
+The protocol is the execution of 
+
+#### Preparation Stage
+
+1. Owner should issue proper number of tokens on Binance Chain as the Pegged 
+2. Owner should create one address on Binance Chain (as OB in the below diagram), and transfer in enough number of tokens for swap
```
Comment (huangsuyu) [inline BEP3.md]: where is the diagram?

File: BEP3.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+2. Via the public balance of OB and APS contract, everyone can check the total pegged token swapped, and ensure the total supply stays clear in the same way as if the token is not pegged
+3. Client gets guarantee on their fund safety. They don’t need to worry too much about the availability of the Deputy process, or any cheat from issuers or swap service providers.
+4. It is possible for one set of infrastructure to handle swap for multiple tokens.
+
+#### Client Swap Tokens from Ethereum to Binance Chain
+
+1. Client calls APS contract, with the hash of a secret, X tokens and a time span T parameters, to express his/her interest to swap X tokens. If the parameters are good (e.g. T>MinLockTime, enough tokens to swap), the call transaction is recorded on the blockchain.
+2. Deputy process monitors the events of the APS contract on Ethereum. If it detects the client’s call and verify good, it will sign and broadcast the HTLT transaction on Binance Chain. This will lock X (or more as bonus) number of pegged tokens on Binance Chain. Please note the time span used in this HTLT transaction should be smaller enough than T in Client’s APS call.
```
Comment (huangsuyu) [inline BEP3.md]: we need to check if the contract has enough deposit

Comment (HaoyangLiu) [reply BEP3.md]: There is no need to check. The contract will ensure there has been enough deposit on the contract address.

**Comment by wjmelements** [comment]
Will it be possible to lock the operator's balance so they can only send using HTLC?

File: BEP3.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+| Random Number Hash | 32 bytes   | hash of a random number, based on SHA256 | No |
+| Coins | []Coin   |similar to the Coins in the original Transfer defined in BEP2| No | 
+| TimeSpan | int64   | number of blocks to wait before the asset may be returned to From if not claimed via Random. The number must be larger than or equal to 360 (>2 minutes), and smaller than 518400 (< 48 hours) | No |
+
+Before the above `To` claims the transferred Coins with the correct random number that can generate the same Random Hash, the Coins will not appear as balance on `To` address. The transaction is signed by private key for `From` address.
+
+
+### Claim Hash Timer Locked 
+
+Claim Hash Timer Locked (CHTL) is to claim the locked asset by showing the Random Number value that matches the hash. Each HTLT locked asset is guaranteed to be release once.
```
Comment (wjmelements) [inline BEP3.md]: Should this be CHLT? This is the only place it is called CHTL.

File: BEP3.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+Here new transaction and logics are required on Binance Chain, as it doesn’t support Smart Contract. Also a standard infrastructure and procedure should be proposed and used as best practice for such inter-chain communications.
+
+The BEP is to tackle the problem, and try to pave the broadway for a practical cross chain solution to empower the decentralized economy with Binance DEX.
+
+# Specification
+
+## Atomic Peg Swap
+
+The primary purpose of HTLC in Binance Chain is to support Atomic Peg Swap (APS). APS is proposed here as the standard way for asset/token issuers to peg or migrate part of the asset/token onto Binance Chain, so that users of the tokens can enjoy the fast transactions and pleasant trading experience.
```
Comment (caffeinum) [inline BEP3.md]: I believe Atomic Swap and Peg should be a totally separate concerns. Atomic Swap is an option, and cannot be enforced without intermediary and complicated mechanisms like DAI (which isn't safe, too).

It's easy to implement an Atomic Swap using HTLC first, so I think we should focus on that and move Peg discussion to separate BEP.

Comment (caffeinum) [reply BEP3.md]: Related:

https://blog.bitmex.com/atomic-swaps-and-distributed-exchanges-the-inadvertent-call-option/

File: BEP3.md (hunk 5 above)
Comment (Dominator008) [inline BEP3.md]: Is "`OB`" an ordinary address controlled by the owner's private key? If so, isn't the owner free to move tokens from this address?

File: BEP3.md (hunk 7 above)
Comment (HaoyangLiu) [reply BEP3.md]: `CHTL` is right.

File: BEP3.md (hunk 5 above)
Comment (HaoyangLiu) [reply BEP3.md]: Yes. It is an ordinary address. But once HTLT transaction is sent, a certain number of coins will be transferred to a temporary address. There is no risk for swap.

Comment (Dominator008) [reply BEP3.md]: Understood the swap process works. My question is geared more towards ensuring the total supply of tokens. It's unclear to me how the current proposal guarantees at any given time, supply(BEP-2) + supply(ERC-20) == initialSupply(ERC-20)?

Comment (HaoyangLiu) [reply BEP3.md]: Taking ERC20 token for instance, the total supply of token A  is 10000. If 5000 the BEP2 token is minted, then 5000 ERC20 token should be burned. Currently, there are 5000 BEP2 token and 5000 ERC20 token. Then the swap service provider use 1000 BEP2 token and 1000 ERC20 token for atomic swap service. Then total supply is still 10000.

Comment (Dominator008) [reply BEP3.md]: But there's no prevention against the token owner minting more BEP-2, besides community monitoring?

**Comment by HaoyangLiu** [comment]
Changes on Binance Chain implementation:
1. Add swap in amount and swap out amount.
2. Add refund msg type on Binance Chain
3. Once a CHTL transaction or a refund transaction are sent, the corresponding hash lock record will be deleted.
4. Change Coins to Coin
5. secetHashLock = hash(secretKey+timestamp), The gap between `timestamp` and block time should be less than one day.

Changes on swap contract interface:
1. Add swap amount in and swap amount out.
2. _timelock should be block height instead of timestamp which is counted by second.
3. Add swap query interface to get all opened swap

I have add the above changes. Please review it again.
@darren-liu @notatestuser @rickyyangz @yutianwu @guagualvcha @ackratos

File: BEP3.md (hunk 5 above)
Comment (HaoyangLiu) [reply BEP3.md]: You are right. The community monitor is the main insurance for the unexpected minting on Binance Chain.

Comment (caffeinum) [reply BEP3.md]: @HaoyangLiu I believe it cannot be called a peg then.

I will state again that I think it's needed to separate Atomic swap BEP and Peg BEP.

**Comment by HaoyangLiu** [comment]
@caffeinum 
Could you clarify how to improve pegging mechanism?

**Comment by Dominator008** [comment]
Without loss of generality, let's design a pegging scheme between ERC-20 and BEP-2. IMO a proper mechanism should be something as follows:

Create pegged BEP-20 tokens (convert a part of the supply from ERC-20 to BEP-2):
1. On Ethereum, the token owner sends a certain amount of ERC-20 tokens to a bridge contract where the tokens are locked until unpegged.
2. On Binance Chain, the token owner broadcasts a `peg` transaction on the Binance Chain with the requested pegging amount.
3. Binance Chain validators check on Ethereum that the locked amount is enough for the pegging and approves the `peg` transaction. Note that they will need to wait for a certain number of Ethereum blocks to establish finality.
4. Token owner is now free to transfer the newly minted pegged tokens on Binance chain.

Destroy pegged BEP-20 tokens (convert a part of they supply from BEP-2 to ERC-20):
1. On Binance Chain, token owner broadcasts a regular `burn` transaction on the Binance chain, which burns certain number of pegged tokens.
2. On Binance Chain, token owner broadcasts a `requestProofOfBurn` transaction, for which the validators sign-off the burnt amount and a nonce with `secp256k1`.
3. On Ethereum, token owner sends the proof of burn to the bridge contract and unlocks the signed-off amount of tokens. The nonce is recorded as used to prevent replay attacks.

Note that this requires an extension to the semantics of `mintable` on Binance chain -- pegged assets should only be mintable with the special `peg` transaction.

This mechanism guarantees that the total supply of the token remains constant.

**Comment by Dominator008** [comment]
https://github.com/swishlabsco/cosmos-ethereum-bridge is example of a unidirectional pegging implementation.

**Comment by karzak** [comment]
> https://github.com/swishlabsco/cosmos-ethereum-bridge is an example of a unidirectional pegging implementation.

FYI The latest version of the ETH bridge code is  https://github.com/musnit/peggy/tree/unidirectional-complete

**Comment by darren-liu** [comment]
> Without loss of generality, let's design a pegging scheme between ERC-20 and BEP-2. IMO a proper mechanism should be something as follows:
> 
> Create pegged BEP-20 tokens (convert a part of the supply from ERC-20 to BEP-2):
> 
> 1. On Ethereum, the token owner sends a certain amount of ERC-20 tokens to a bridge contract where the tokens are locked until unpegged.
> 2. On Binance Chain, the token owner broadcasts a `peg` transaction with the requested pegging amount.
> 3. Binance Chain validators check on Ethereum that the locked amount is enough for the pegging and approves the `peg` transaction. Note that they will need to wait for a certain number of Ethereum blocks to establish finality.
> 4. Token owner is now free to transfer the newly minted pegged tokens on Binance Chain.
> 
> Destroy pegged BEP-20 tokens (convert a part of they supply from BEP-2 to ERC-20):
> 
> 1. On Binance Chain, token owner broadcasts a regular `burn` transaction on the Binance chain, which burns certain number of pegged tokens.
> 2. On Binance Chain, token owner broadcasts a `requestProofOfBurn` transaction, for which the validators sign-off the burnt amount and a nonce with `secp256k1`.
> 3. On Ethereum, token owner sends the proof of burn to the bridge contract. The contract checks the signatures of the validators and unlocks the signed-off amount of tokens. The nonce is recorded as used to prevent replay attacks.
> 
> Note that this requires an extension to the semantics of `mintable` on Binance chain -- pegged assets should only be mintable with the special `peg` transaction.
> 
> This mechanism guarantees that the total supply of the token remains constant.

It is nice to have it trustless, but it is heavy-weight-lifting and I don't think it is achievable to implement this feature for multiple chains.

**Comment by Dominator008** [comment]
> > Without loss of generality, let's design a pegging scheme between ERC-20 and BEP-2. IMO a proper mechanism should be something as follows:
> > Create pegged BEP-20 tokens (convert a part of the supply from ERC-20 to BEP-2):
> > 
> > 1. On Ethereum, the token owner sends a certain amount of ERC-20 tokens to a bridge contract where the tokens are locked until unpegged.
> > 2. On Binance Chain, the token owner broadcasts a `peg` transaction with the requested pegging amount.
> > 3. Binance Chain validators check on Ethereum that the locked amount is enough for the pegging and approves the `peg` transaction. Note that they will need to wait for a certain number of Ethereum blocks to establish finality.
> > 4. Token owner is now free to transfer the newly minted pegged tokens on Binance Chain.
> > 
> > Destroy pegged BEP-20 tokens (convert a part of they supply from BEP-2 to ERC-20):
> > 
> > 1. On Binance Chain, token owner broadcasts a regular `burn` transaction on the Binance chain, which burns certain number of pegged tokens.
> > 2. On Binance Chain, token owner broadcasts a `requestProofOfBurn` transaction, for which the validators sign-off the burnt amount and a nonce with `secp256k1`.
> > 3. On Ethereum, token owner sends the proof of burn to the bridge contract. The contract checks the signatures of the validators and unlocks the signed-off amount of tokens. The nonce is recorded as used to prevent replay attacks.
> > 
> > Note that this requires an extension to the semantics of `mintable` on Binance chain -- pegged assets should only be mintable with the special `peg` transaction.
> > This mechanism guarantees that the total supply of the token remains constant.
> 
> It is nice to have it trustless, but it is heavy-weight-lifting and I don't think it is realistically achievable to implement this feature for multiple chains.
> 
> I agree it is not a perfect solution for Peg here. The progress is to make at least a better way than current situation of how peg is performed on Binance Chain.

Understood. We can revisit once projects like ThorChain have made more progress with generic peg zone implementations.

**Pull request merged**